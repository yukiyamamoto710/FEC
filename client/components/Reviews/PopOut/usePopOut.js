import { useState, useEffect } from 'react';
import postReview from '../func/postReview/postReview';

const modle = {
  product_id: '',
  summary: '',
  name: '',
  email: '',
};

export default function usePopOut({
  key, obj, data, addUserReview, cancelAddReview,
}) {
  const [userReview, setUserReview] = useState(modle);
  const [recommend, setRecommend] = useState(true);
  const [photos, setPhotos] = useState([]);
  const [rating, setRating] = useState(0);
  const [body, setBody] = useState('');
  const [characteristics, setCharacteristics] = useState({});
  const [listWarning, setListWarning] = useState([]);

  useEffect(() => {
    setCharacteristics(obj);
    const modle1 = modle;
    modle1.product_id = data.product_id;
    setUserReview(modle1);
    setRecommend(true);
    setPhotos([]);
    setBody([]);
    setRating(0);
    setListWarning([]);
  }, [data]);

  useEffect(() => {
    setUserReview(userReview);
  }, [userReview]);

  const starClick = (event) => {
    const newObj = { ...characteristics };
    const arr = event.target.id.split(' ');
    const newKey = arr[0];
    const value = arr[1];
    if (newKey === 'Stars') {
      setRating(Number(value));
    } else {
      newObj[newKey].value = String(value);
      setCharacteristics(newObj);
    }
  };

  const isShowWarning = () => {
    setListWarning([]);
  };

  const handleClickCheckReview = () => {
    const {
      email,
      name,
      summary,
    } = userReview;
    const warning = [];
    if (body.length < 50) { warning.push('Description'); }
    if (!email.includes('@') || !email.includes('.')) {
      warning.push('Email');
    } else if (email.indexOf('@') >= email.indexOf('.') - 1) {
      warning.push('Email');
    } else if (email.indexOf('@') === 0 || email.indexOf('.') === email.length-1) {
      warning.push('Email');
    }
    if (name.length === 0) { warning.push('Name'); }
    if (summary.length === 0) { warning.push('Title'); }
    if (rating === 0) { warning.push('Stars'); }
    for (let i = 0; i < key.length; i += 1) {
      if (!characteristics[key[i]].value) {
        warning.push(key[i]);
      }
    }
    if (warning.length > 0) {
      setListWarning(warning);
    } else {
      const newObj = { ...userReview };
      newObj.product_id = Number(userReview.product_id);
      newObj.body = body;
      newObj.rating = rating;
      newObj.photos = [...photos];
      newObj.recommend = recommend;
      newObj.characteristics = { };
      const key2 = Object.keys(characteristics);
      for (let i = 0; i < key2.length; i += 1) {
        const c = characteristics;
        newObj.characteristics[c[key2[i]].id] = Number(c[key2[i]].value);
      }

      cancelAddReview();
      addUserReview(newObj);
      postReview(newObj, userReview.product_id);
    }
  };

  const handleChangeTarget = (event) => {
    const { id, value } = event.target;
    if (id === 'recommend') {
      if (value === 'false') {
        setRecommend(false);
      } else {
        setRecommend(true);
      }
    } else if (id === 'photos') {
      if (value !== undefined) {
        setPhotos([...photos, String(value)]);
      }
    } else if (id === 'body') {
      setBody(value);
    } else {
      userReview[id] = value;
    }
  };
  return {
    handleChangeTarget,
    handleClickCheckReview,
    isShowWarning,
    starClick,
    userReview,
    recommend,
    photos,
    rating,
    body,
    characteristics,
    listWarning,
  };
}
