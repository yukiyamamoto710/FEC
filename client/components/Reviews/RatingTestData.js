const testData1 = {
  characteristics: {
    Comfort: {
      id: 86687,
      value: '2.7500000000000000',
    },
    Quality: {
      id: 86688,
      value: '2.0000000000000000',
    },
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
    Width: {
      id: 86686,
      value: '3.2500000000000000',
    },
  },
  product_id: '25821',
  ratings: {
    1: '1',
    4: '1',
    5: '2',
  },
  recommended: {
    false: '1',
    true: '3',
  },
};
const testData2 = {
  characteristics: {
    Size: {
      id: 86685,
      value: '2.2500000000000000',
    },
  },
  product_id: '25721',
  ratings: {
  },
  recommended: {
  },
};

export { testData1, testData2 };
