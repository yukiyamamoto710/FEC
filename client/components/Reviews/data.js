const Data = {
  Size: [
    'A size too small',
    '½ a size too small',
    'Perfect',
    '½ a size too big',
    'A size too wide',
  ],
  Width: [
    'Too narrow',
    'Slightly narrow',
    'Perfect',
    'Slightly wide',
    'Too wide',
  ],
  Comfort: [
    'Uncomfortable',
    'Slightly uncomfortable',
    'Ok',
    'Comfortable',
    'Perfect',
  ],
  Quality: [
    'Poor',
    'Below average',
    'What I expected',
    'Pretty great',
    'Perfect',
  ],
  Length: [
    'Runs Short',
    'Runs slightly short',
    'Perfect',
    'Runs slightly long',
    'Runs long',
  ],
  Fit: [
    'Runs tight',
    'Runs slightly tight',
    'Perfect',
    'Runs slightly long',
    'Runs long',
  ],
  Stars: [
    'Poor',
    'Fair',
    'Average',
    'Good',
    'Great',
  ],
};

const msg = [
  'For privacy reasons,do not use your full name or email address',
  'For authentication reasons, you will not be emailed',
];

const textExample = [
  'Example: jackson11!',
  'jackson11@email.com',
  'Example: Best purchase ever!',
  'Why did you like the product or not? and need a counter min 50 max 1000',
];

export { Data, msg, textExample };
