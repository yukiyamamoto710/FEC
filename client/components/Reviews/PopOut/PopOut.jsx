import React from 'react';
import PropTypes from 'prop-types';

import FormStarCK from './FormStarCK/FormStarCK';
import { msg, textExample } from '../data';
import Warning from './Warning/Warning';
import usePopOut from './usePopOut';

const PopOut = (props) => {
  const { addUserReview, data, cancelAddReview } = props;
  const key = Object.keys(data.characteristics);
  const obj = {};
  for (let i = 0; i < key.length; i += 1) {
    obj[key[i]] = {
      id: data.characteristics[key[i]].id,
      value: 0,
    };
  }
  const {
    handleChangeTarget,
    handleClickCheckReview,
    isShowWarning,
    starClick,
    userReview,
    recommend,
    rating,
    body,
    characteristics,
    listWarning,
  } = usePopOut({
    key, obj, data, addUserReview, cancelAddReview,
  });

  return (
    <div
      className="popOut"
      data-testid="popout"
    >
      {listWarning.length !== 0
        ? (
          <Warning
            data-testid="warning"
            listWarning={listWarning}
            isShowWarning={isShowWarning}
          />
        ) : null}
      <div className="popOutContainer">
        <div className="topContainer">
          <div className="topSmallContainer">
            <div>Recommend</div>
            <div className="buttonBase">
              <button
                data-testid="recommendYes"
                type="button"
                onClick={handleChangeTarget}
                className={recommend
                  ? 'clickedButton'
                  : 'unclickedButton'}
                id="recommend"
                value="true"
              >
                YES
              </button>
              <button
                data-testid="recommendNo"
                type="button"
                onClick={handleChangeTarget}
                className={recommend
                  ? 'unclickedButton'
                  : 'clickedButton'}
                id="recommend"
                value={false}
              >
                No
              </button>
            </div>
          </div>
          <button
            className="formButton"
            type="button"
            onClick={cancelAddReview}
          >
            X
          </button>
        </div>
        <div className="firstContainer">
          <div className="firstTopContainer">
            <FormStarCK
              starClick={starClick}
              info={Number(rating)}
              name="Stars"
            />
            {Object.keys(characteristics).map((i) => (
              <FormStarCK
                starClick={starClick}
                key={i}
                info={Number(characteristics[i].value)}
                name={i}
              />
            ))}
          </div>
          <div className="midSmallContainer">
            <div className="tagContainer">
              <div className="tag">Name :</div>
              <div className="tag">Email :</div>
              <div className="tag">Title :</div>
              <div className="tag">Description :</div>
              <div className="tag">
                Count :
                {body.length}
              </div>
            </div>
            <div className="textcontainer">
              { ['name', 'email', 'summary'].map((i, index) => (
                <div key={i} className="formTextContainer">
                  <div>
                    <input
                      data-testid={`formText${index}`}
                      className="formText"
                      type="text"
                      maxLength="60"
                      onChange={handleChangeTarget}
                      placeholder={textExample[index]}
                      id={i}
                    />
                    {userReview[i].length === 0 || msg[index] === undefined
                      ? <div className="block" />
                      : <div className="formMSG">{msg[index]}</div>}
                  </div>
                </div>
              ))}
              <textarea
                data-testid="body"
                id="body"
                className="textarea"
                rows="5"
                maxLength="1000"
                onChange={handleChangeTarget}
                placeholder={textExample[3]}
              />
            </div>
          </div>
        </div>
        <div className="btnContainer">
          <button
            type="button"
            data-testid="submitBtn"
            className="clickedButton"
            onClick={handleClickCheckReview}
          >
            Submit
          </button>
          <input
            className="clickedButton"
            type="file"
            key="photo1"
            onChange={handleChangeTarget}
            id="photos"
          />
          <input
            className="clickedButton"
            type="file"
            key="photo2"
            onChange={handleChangeTarget}
            id="photos"
          />
        </div>
      </div>
    </div>
  );
};

PopOut.propTypes = {
  data: PropTypes.shape({
    characteristics: PropTypes.shape({}),
    product_id: PropTypes.string,
  }),
  addUserReview: PropTypes.func,
  cancelAddReview: PropTypes.func,
};
PopOut.defaultProps = {
  data: { characteristics: {}, product_id: '25711' },
  addUserReview: () => 1,
  cancelAddReview: () => 1,
};
export default PopOut;
