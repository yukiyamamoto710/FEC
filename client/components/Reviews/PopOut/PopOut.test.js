/**
 * @jest-environment jsdom
 */
import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  cleanup,
  fireEvent,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import PopOut from './PopOut';
import { testData1 } from '../RatingTestData';
import { msg } from '../data';

afterEach(cleanup);

it('render correctly', () => {
  const { getByTestId } = render(<PopOut data={testData1} />);
  expect(getByTestId('popout')).toBeInTheDocument();
});

it('render correctly', () => {
  const { getByTestId } = render(<PopOut data={testData1} />);
  fireEvent.click(getByTestId('submitBtn'));
  expect(getByTestId('WarningContainer')).toBeInTheDocument();
});

it('render correctly', () => {
  const { getByTestId } = render(<PopOut data={testData1} />);
  fireEvent.click(getByTestId('recommendNo'));
  expect(getByTestId('recommendNo').classList.contains('clickedButton')).toBe(true);
});

it('render correctly', () => {
  const { getByTestId, queryByText } = render(<PopOut data={testData1} />);
  fireEvent.change(getByTestId('formText1'), { target: { value: 'd' } });
  fireEvent.change(getByTestId('body'), { target: { value: 'd' } });
  expect(queryByText(msg[1])).toBeInTheDocument();
});

it('should get default addUserReview and cancelAddReview', () => {
  expect(PopOut.defaultProps.addUserReview()).toBeDefined();
  expect(PopOut.defaultProps.cancelAddReview()).toBeDefined();
});

it('matches snapShot', () => {
  const tree = renderer.create(
    <PopOut
      data={testData1}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
