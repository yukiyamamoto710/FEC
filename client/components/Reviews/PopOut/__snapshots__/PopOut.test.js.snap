// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<div
  className="popOut"
  data-testid="popout"
>
  <div
    className="popOutContainer"
  >
    <div
      className="topContainer"
    >
      <div
        className="topSmallContainer"
      >
        <div>
          Recommend
        </div>
        <div
          className="buttonBase"
        >
          <button
            className="clickedButton"
            data-testid="recommendYes"
            id="recommend"
            onClick={[Function]}
            type="button"
            value="true"
          >
            YES
          </button>
          <button
            className="unclickedButton"
            data-testid="recommendNo"
            id="recommend"
            onClick={[Function]}
            type="button"
            value={false}
          >
            No
          </button>
        </div>
      </div>
      <button
        className="formButton"
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </div>
    <div
      className="firstContainer"
    >
      <div
        className="firstTopContainer"
      >
        <div
          className="topSmallContainer"
          data-testid="formStarCK"
        >
          <div>
            Stars
          </div>
          <div
            className="ratingHead"
            data-testid="stars"
          >
            <div>
              <div
                className="formStar emptyStar"
                data-testid="button0"
                id="Stars 1"
                label="emptyStar"
                name="Stars"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              />
            </div>
            <div>
              <div
                className="formStar emptyStar"
                data-testid="button1"
                id="Stars 2"
                label="emptyStar"
                name="Stars"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              />
            </div>
            <div>
              <div
                className="formStar emptyStar"
                data-testid="button2"
                id="Stars 3"
                label="emptyStar"
                name="Stars"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              />
            </div>
            <div>
              <div
                className="formStar emptyStar"
                data-testid="button3"
                id="Stars 4"
                label="emptyStar"
                name="Stars"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              />
            </div>
            <div>
              <div
                className="formStar emptyStar"
                data-testid="button4"
                id="Stars 5"
                label="emptyStar"
                name="Stars"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              />
            </div>
          </div>
          <div
            className="formText"
            data-testid="formStarCKText"
          />
        </div>
      </div>
      <div
        className="midSmallContainer"
      >
        <div
          className="tagContainer"
        >
          <div
            className="tag"
          >
            Name :
          </div>
          <div
            className="tag"
          >
            Email :
          </div>
          <div
            className="tag"
          >
            Title :
          </div>
          <div
            className="tag"
          >
            Description :
          </div>
          <div
            className="tag"
          >
            Count :
            0
          </div>
        </div>
        <div
          className="textcontainer"
        >
          <div
            className="formTextContainer"
          >
            <div>
              <input
                className="formText"
                data-testid="formText0"
                id="name"
                maxLength="60"
                onChange={[Function]}
                placeholder="Example: jackson11!"
                type="text"
              />
              <div
                className="block"
              />
            </div>
          </div>
          <div
            className="formTextContainer"
          >
            <div>
              <input
                className="formText"
                data-testid="formText1"
                id="email"
                maxLength="60"
                onChange={[Function]}
                placeholder="jackson11@email.com"
                type="text"
              />
              <div
                className="formMSG"
              >
                For authentication reasons, you will not be emailed
              </div>
            </div>
          </div>
          <div
            className="formTextContainer"
          >
            <div>
              <input
                className="formText"
                data-testid="formText2"
                id="summary"
                maxLength="60"
                onChange={[Function]}
                placeholder="Example: Best purchase ever!"
                type="text"
              />
              <div
                className="block"
              />
            </div>
          </div>
          <textarea
            className="textarea"
            data-testid="body"
            id="body"
            maxLength="1000"
            onChange={[Function]}
            placeholder="Why did you like the product or not? and need a counter min 50 max 1000"
            rows="5"
          />
        </div>
      </div>
    </div>
    <div
      className="btnContainer"
    >
      <button
        className="clickedButton"
        data-testid="submitBtn"
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
      <input
        className="clickedButton"
        id="photos"
        onChange={[Function]}
        type="file"
      />
      <input
        className="clickedButton"
        id="photos"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
</div>
`;
